app/server/src/routes/userProfile/userProfile.controller.js

```diff
@@ -37,23 +48,104 @@ const UserProfileController = {
+      if(user.followed_users.includes(id)){// already following
+        return res.status(400).json({ message: "You already follow this user!" });
+      }
+      user.followed_users.push(other_user);
+      other_user.follower_users.push(user);
+      await user.save();
+      await other_user.save();
+      return res.status(200).json({ message: `${user.name+" "+user.surname} started following ${other_user.name+" "+ other_user.surname}` });
+    } catch (e) {
+      return res.status(400).json({ error: e.toString() });
+    }
+  },
+
+  unfollow: async function (req, res) {
+    try {
+      const id = req.body.user_id;
+      const user_id = req.auth.id;
+      if(id == user_id.toString()){
+        return res.status(400).json({ message: "You can't unfollow yourself!" });
+      }
+      const user = await UserModel.User.findById(user_id);
+      const other_user = await UserModel.User.findById(id);
+      if(!(user.followed_users.includes(id))){// already following
+        return res.status(400).json({ message: "You are already not following this user!" });
+      }
+      let index = user.followed_users.indexOf(id);
+      if (index > -1) { // only splice array when item is found
+        user.followed_users.splice(index, 1); // 2nd parameter means remove one item only
+      }
+      index = other_user.follower_users.indexOf(user_id.toString());
+      if (index > -1) { // only splice array when item is found
+        other_user.follower_users.splice(index, 1); // 2nd parameter means remove one item only
+      }
+      await user.save();
+      await other_user.save();
+      return res.status(200).json({ message: `${user.name+" "+user.surname} stopped following ${other_user.name+" "+ other_user.surname}` });
+    } catch (e) {
+      return res.status(400).json({ error: e.toString() });
     }
   },
 
   getProfile: async function (req, res) {
     try {
       const id = req.params.id;
-      // Need a new function in user model that returns populated versions of personal info and badges
-      const profile = await UserModel.getPopulatedPersonalInfo(id);
-      res.status(200).json({ profile });
+      // if user logged-in
+      if (req.auth) {
+        var user_id = req.auth.id;
+        const user = await UserModel.User.findById(user_id);
+        if(id == user_id.toString()){// own profile
+          profile = await UserModel.getPopulatedPersonalInfo(id);
+          return res.status(200).json({ profile });
+        }
+        if(user.followed_users.includes(id)){// looking at the profile of a user we follow
+          profile = await UserModel.getPopulatedPersonalInfo(id);
+          return res.status(200).json({ profile });
+        }
+        // looking at a profile we don't follow
+        const other_user = await UserModel.getUserByID(id);
+        let profile;
+        if(other_user.is_private){
+          profile = await UserModel.getPopulatedPersonalInfoPrivate(id);
+        }
```

Comment:
Should we show the email and the picture of a private profile?

Reply:
Yes, small image will be shown, email may be removed.
![image](https://user-images.githubusercontent.com/49765256/205458531-facd03c6-a09d-4a77-b9f6-4ad3e15e16bc.png)